src/shell/hooks/useLocalStorage.ts

```diff
@@ -0,0 +1,107 @@
+/*
+ * SPDX-FileCopyrightText: 2023 Zextras <https://www.zextras.com>
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+import React, { useCallback, useEffect, useMemo, useRef } from 'react';
+import { create } from 'zustand';
+
+function isSameLocalStorageValue(valueA: unknown, valueB: unknown): boolean {
+	return JSON.stringify(valueA) === JSON.stringify(valueB);
+}
+
+type LocalStorageOptions = { keepSynchedWithStorage?: boolean };
+
+type LocalStorageState = {
+	storage: Record<string, any>;
+	readValue: <T>(key: string, fallback: T) => void;
+	setValue: <T>(key: string, value: React.SetStateAction<T>) => void;
+};
+
+const useLocalStorageStore = create<LocalStorageState>()((setState) => ({
+	storage: {},
+	readValue<T>(key: string, fallback: T): void {
+		try {
+			const localStorageItem = window.localStorage.getItem(key);
+			const item = localStorageItem !== null ? JSON.parse(localStorageItem) : fallback;
+			setState((state) => {
+				if (!isSameLocalStorageValue(item, state.storage[key])) {
+					const newState = { ...state };
+					newState.storage[key] = item;
+					return newState;
+				}
+				return state;
+			});
+		} catch (error) {
+			console.error(error);
+			setState((state) => {
+				const newState = { ...state };
+				newState.storage[key] = fallback;
+				return newState;
+			});
+		}
+	},
+	setValue<T>(key: string, value: React.SetStateAction<T>): void {
+		setState((state) => {
+			const valueToStore = value instanceof Function ? value(state.storage[key]) : value;
+			const newState = { ...state };
+			if (!isSameLocalStorageValue(valueToStore, state.storage[key])) {
+				window.localStorage.setItem(key, JSON.stringify(valueToStore));
+				newState.storage[key] = value;
```

Comment:
It is wrongly saving the function instead of the value inside the store if the set is called with the (prevState) => state 
```suggestion
				newState.storage[key] = valueToStore;
```